src/components/heartrate/Ppg.cpp

```diff
@@ -2,98 +2,242 @@
-      auto d = d1[i] - d2[i];
-      e += d * d;
-    }
-    return e;
-  }
+// Refactored for frequency based algorithm - Ceimour
+Ppg::Ppg() {
+  dataAverage.fill(0.0f);
+  spectrum.fill(0.0f);
+}
 
-  int CompareShift(int8_t* d, int shift, size_t count) {
-    return Compare(d + shift, d, count - shift);
+int8_t Ppg::Preprocess(float hrs, uint32_t als) {
+  if (dataIndex < dataLength) {
+    dataHRS[dataIndex++] = hrs;
   }
+  alsValue = als;
+  if (alsValue > alsThreshold)
+    return 1;
+  else
+    return 0;
+}
 
-  int Trough(int8_t* d, size_t size, uint8_t mn, uint8_t mx) {
-    auto z2 = CompareShift(d, mn - 2, size);
-    auto z1 = CompareShift(d, mn - 1, size);
-    for (int i = mn; i < mx + 1; i++) {
-      auto z = CompareShift(d, i, size);
-      if (z2 > z1 && z1 < z)
-        return i;
-      z2 = z1;
-      z1 = z;
+int Ppg::HeartRate() {
+  if (dataIndex < dataLength)
+    return 0;
+
+  int hr = 0;
+  hr = ProcessHeartRate(resetSpectralAvg);
+  resetSpectralAvg = false;
+  // Make room for overlapWindow number of new samples
+  for (int i = 0; i < dataLength - overlapWindow; i++) {
+    dataHRS[i] = dataHRS[i + overlapWindow];
+  }
+  dataIndex = dataLength - overlapWindow;
+  return hr;
+}
+// Pass init == true to reset spectral averaging.
+// Returns -1 (Reset Acquisition), 0 (Unable to obtain HR) or HR (BPM).
+int Ppg::ProcessHeartRate(bool init) {
+  std::copy(dataHRS.begin(), dataHRS.end(), vReal.begin());
+  Detrend(vReal.data(), dataLength);
+  Filter30to240(vReal.data(), dataLength);
+  // Compute in place power spectrum
+  vImag.fill(0.0f);
+  FFT.Windowing(vReal.data(), static_cast<uint16_t>(dataLength), FFT_WIN_TYP_HANN, FFT_FORWARD);
+  FFT.Compute(vReal.data(), vImag.data(), static_cast<uint16_t>(dataLength), FFT_FORWARD);
+  FFT.ComplexToMagnitude(vReal.data(), vImag.data(), static_cast<uint16_t>(dataLength));
+  SpectrumAverage(vReal.data(), spectrum.data(), spectrum.size(), init);
+  peakLocation = 0.0f;
+  float threshold = peakDetectionThreshold;
+  float peakWidth = 0.0f;
+  int specLen = spectrum.size();
+  float max = Max(spectrum.data(), hrROIbegin, hrROIend);
+  float signalToNoiseRatio = SignalToNoise(spectrum.data(), max, hrROIbegin, hrROIend);
+  if (signalToNoiseRatio > signalToNoiseThreshold && spectrum.at(0) < dcThreshold) {
+    threshold *= max;
+    // Reuse VImag for interpolation x values passed to PeakSearch
+    for (int i = 0; i < dataLength; i++) {
+      vImag[i] = i;
     }
+    peakLocation = PeakSearch(vImag.data(),
+                              spectrum.data(),
+                              specLen,
+                              static_cast<float>(hrROIbegin),
+                              static_cast<float>(hrROIend),
+                              threshold,
+                              peakWidth);
+    peakLocation *= freqResolution;
+  }
+  // Peak too wide? (broad spectrum noise or large, rapid HR change)
+  if (peakWidth > maxPeakWidth)
+    peakLocation = 0.0f;
```

Comment:
Always use braces.

Reply:
These should be fixed in recent commits.